x-pack/solutions/observability/plugins/apm/server/routes/assistant_functions/get_red_metrics/index.ts

```diff
@@ -0,0 +1,108 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { kqlQuery, rangeQuery } from '@kbn/observability-plugin/server';
+import { SERVICE_NAME } from '../../../../common/es_fields/apm';
+import { ApmDocumentType } from '../../../../common/document_type';
+import { RollupInterval } from '../../../../common/rollup';
+import type { APMEventClient } from '../../../lib/helpers/create_es_client/create_apm_event_client';
+import {
+  calculateFailedTransactionRate,
+  getOutcomeAggregation,
+} from '../../../lib/helpers/transaction_error_rate';
+import { calculateThroughputWithRange } from '../../../lib/helpers/calculate_throughput';
+import { getDurationFieldForTransactions } from '../../../lib/helpers/transactions';
+
+const MAX_NUMBER_OF_GROUPS = 100;
+
+export interface RedMetricsItem {
+  group: string;
+  latency: number | null;
+  throughput: number;
+  failureRate: number;
+}
+
+export interface GetRedMetricsResponse {
+  items: RedMetricsItem[];
+}
+
+export async function getRedMetrics({
+  apmEventClient,
+  start,
+  end,
+  filter,
+  groupBy = SERVICE_NAME,
+}: {
+  apmEventClient: APMEventClient;
+  start: number;
+  end: number;
+  filter?: string;
+  groupBy?: string;
+}): Promise<GetRedMetricsResponse> {
+  const documentType = ApmDocumentType.TransactionMetric;
+  const rollupInterval = RollupInterval.OneMinute;
+  const durationField = getDurationFieldForTransactions(documentType, true);
+
+  const outcomeAggs = getOutcomeAggregation(documentType);
+
+  const response = await apmEventClient.search('get_service_red_metrics', {
+    apm: {
+      sources: [{ documentType, rollupInterval }],
+    },
+    size: 0,
+    track_total_hits: false,
+    query: {
+      bool: {
+        filter: [...rangeQuery(start, end), ...kqlQuery(filter)],
+      },
+    },
+    aggs: {
+      groups: {
+        terms: {
+          field: groupBy,
+          size: MAX_NUMBER_OF_GROUPS,
+          order: { _count: 'desc' },
+        },
+        aggs: {
+          avg_latency: {
+            avg: {
+              field: durationField,
+            },
+          },
+          ...outcomeAggs,
+        },
+      },
+    },
+  });
+
+  const buckets = response.aggregations?.groups?.buckets ?? [];
+
+  const items: RedMetricsItem[] = buckets.map((bucket) => {
+    const docCount = bucket.doc_count;
+    const latencyValue = bucket.avg_latency?.value;
+
+    const latencyMs =
+      latencyValue !== null && latencyValue !== undefined ? latencyValue / 1000 : null;
+
+    const failureRate = calculateFailedTransactionRate(bucket);
+
+    const throughput = calculateThroughputWithRange({
+      start,
+      end,
+      value: docCount,
+    });
+
+    return {
+      group: bucket.key as string,
+      latency: latencyMs,
+      throughput,
+      failureRate,
+    };
+  });
+
+  return { items };
```

Comment:
nit: why the need for an object?
```suggestion
  return items;
```

Reply:
Updated - https://github.com/elastic/kibana/pull/247474/commits/423900df1b5b3255b766ab4e64a74551cf3f65a2